x/dsmr/node_test.go

```diff
@@ -994,11 +1057,218 @@ func TestNode_NewBlock_IncludesChunkCerts(t *testing.T) {
+		},
+		{
+			name:          "empty block",
+			parentBlocks:  [][]int{{1}, {2}, {3}, {4}},
+			chunks:        []int{},
+			timestamp:     6,
+			verifyWantErr: ErrEmptyBlock,
+			buildWantErr:  ErrNoAvailableChunkCerts,
+		},
+		{
+			name:                       "three blocks non consecutive duplicate chunks outside validity window",
+			parentBlocks:               [][]int{{4}, {5}},
+			chunks:                     []int{4},
+			timestamp:                  4,
+			verifyWantErr:              validitywindow.ErrDuplicateContainer, // this isn't ideal, since it would disqualify a duplicate chunk even when it's outside the validity window. However, it default to the correct direction.
+			buildWantErr:               ErrNoAvailableChunkCerts,
+			validalidityWindowDuration: 1,
+		},
+	}
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
+			validationWindow := testCase.validalidityWindowDuration
+			if validationWindow == 0 {
+				validationWindow = testingDefaultValidityWindowDuration
+			}
+
+			var node *Node[tx]
+			nodes, indexer := newNodes(t, 1)
+			node = nodes[0]
+			node.validityWindowDuration = validationWindow
+
+			// initialize node history.
+			parentBlk := node.LastAccepted
+			for _, chunkList := range testCase.parentBlocks {
+				blk := Block{
+					ParentID:  parentBlk.GetID(),
+					Height:    uint64(int(node.LastAccepted.Height) + 1),
+					Timestamp: int64(int(node.LastAccepted.Timestamp) + 1),
+					blkID:     ids.GenerateTestID(),
+				}
+
+				for _, chunkExpiry := range chunkList {
+					_, chunkCert, err := node.BuildChunk(
+						context.Background(),
+						[]tx{
+							{
+								ID:     ids.Empty,
+								Expiry: int64(chunkExpiry),
+							},
+						},
+						int64(chunkExpiry),
+						codec.Address{},
+					)
+					r.NoError(err)
+					blk.ChunkCerts = append(blk.ChunkCerts, &chunkCert)
+				}
+
+				r.NoError(node.Verify(context.Background(), parentBlk, blk))
+
+				r.NoError(node.Accept(context.Background(), blk))
+				indexer.set(blk.GetID(), NewExecutionBlock(blk))
+				parentBlk = blk
+			}
+
+			// create the block so that we can test it against the execute directly.
+			newBlk := Block{
+				ParentID:  parentBlk.GetID(),
+				Height:    uint64(testCase.timestamp),
+				Timestamp: testCase.timestamp,
+				blkID:     ids.GenerateTestID(),
+			}
+
+			for _, chunkExpiry := range testCase.chunks {
+				_, chunkCert, err := node.BuildChunk(
+					context.Background(),
+					[]tx{
+						{
+							ID:     ids.Empty, // ids.GenerateTestID(),
+							Expiry: int64(chunkExpiry),
+						},
+					},
+					int64(chunkExpiry),
+					codec.Address{},
+				)
+				r.NoError(err)
+				newBlk.ChunkCerts = append(newBlk.ChunkCerts, &chunkCert)
+			}
+			_, err := node.BuildBlock(context.Background(), parentBlk, testCase.timestamp)
+			r.ErrorIs(err, testCase.buildWantErr)
+
+			r.ErrorIs(node.Verify(context.Background(), parentBlk, newBlk), testCase.verifyWantErr)
+		})
```

Comment:
If `BuildBlock` succeeds, it seems that we are expecting it to build the same block that we have produced locally. Should we be checking that they're the same here?

Reply:
if the `BuildBlock` won't return any of the expected errors, than it would return a valid block. However, I'm not sure how easy it would be to compare it to the testing block we've created. I'll add ParentID, Height and Timestamp.